crates/nu-engine/src/closure_eval.rs

```diff
@@ -0,0 +1,236 @@
+/// It is first created via [`ClosureEval::new`],
+/// then has arguments added via [`ClosureEval::add_arg`],
+/// and then can be run using [`ClosureEval::run_with_input`].
+///
+/// ```no_run
+/// # use nu_protocol::{PipelineData, Value};
+/// # use nu_engine::ClosureEval;
+/// # let engine_state = unimplemented!();
+/// # let stack = unimplemented!();
+/// # let closure = unimplemented!();
+/// let mut closure = ClosureEval::new(engine_state, stack, closure);
+/// let iter = Vec::<Value>::new()
+///     .into_iter()
+///     .map(move |value| closure.add_arg(value).run_with_input(PipelineData::Empty));
+/// ```
+///
+/// Many closures follow a simple, common scheme where the pipeline input and the first argument are the same value.
+/// In this case, use [`ClosureEval::run_with_value`]:
+///
+/// ```no_run
+/// # use nu_protocol::{PipelineData, Value};
+/// # use nu_engine::ClosureEval;
+/// # let engine_state = unimplemented!();
+/// # let stack = unimplemented!();
+/// # let closure = unimplemented!();
+/// let mut closure = ClosureEval::new(engine_state, stack, closure);
+/// let iter = Vec::<Value>::new()
+///     .into_iter()
+///     .map(move |value| closure.run_with_value(value));
+/// ```
+///
+/// Environment isolation and other cleanup is handled by [`ClosureEval`],
+/// so nothing needs to be done following [`ClosureEval::run_with_input`] or [`ClosureEval::run_with_value`].
+pub struct ClosureEval {
+    engine_state: EngineState,
+    stack: Stack,
+    block: Arc<Block>,
+    arg_index: usize,
+    env_vars: Vec<EnvVars>,
+    env_hidden: HashMap<String, HashSet<String>>,
+    eval: EvalBlockWithEarlyReturnFn,
+}
+
+impl ClosureEval {
+    /// Create a new [`ClosureEval`].
+    pub fn new(engine_state: &EngineState, stack: &Stack, closure: Closure) -> Self {
+        let engine_state = engine_state.clone();
+        let stack = stack.captures_to_stack(closure.captures);
+        let block = engine_state.get_block(closure.block_id).clone();
+        let env_vars = stack.env_vars.clone();
+        let env_hidden = stack.env_hidden.clone();
+        let eval = get_eval_block_with_early_return(&engine_state);
+
+        Self {
+            engine_state,
+            stack,
+            block,
+            arg_index: 0,
+            env_vars,
+            env_hidden,
+            eval,
+        }
+    }
+
+    /// Sets whether to enable debugging when evaluating the closure.
+    ///
+    /// By default, this is controlled by the [`EngineState`] used to create this [`ClosureEval`].
+    pub fn debug(&mut self, debug: bool) -> &mut Self {
+        self.eval = eval_fn(debug);
+        self
+    }
+
+    fn try_add_arg(&mut self, value: Cow<Value>) {
+        if let Some(var_id) = self
+            .block
+            .signature
+            .get_positional(self.arg_index)
+            .and_then(|var| var.var_id)
+        {
+            self.stack.add_var(var_id, value.into_owned());
+            self.arg_index += 1;
+        }
```

Comment:
Shouldn't it return an error if we try to insert too many arguments? It seems like this would silently not do anything if you keep adding args beyond the closure's number of positionals.

Reply:
We could, but parameters are often optional for the closures, so it may be hard to tell when it is an error or if it is allowed/expected.

Reply:
Ah, ok. Then maybe a check if arg_index doesn't go beyond total number of parameters? Either way, it's OK, just something I wondered.